resource-aggregate/cqrs/eventstore/mongodb/eventstore.go

```diff
@@ -271,234 +299,317 @@ func (s *EventStore) ensureIndex(ctx context.Context, col *mongo.Collection, ind
+// All events must have same AggregateID and GroupID.
+// All events must have filled AggregateID, GroupID and Eventype.
+// Versions among events must continues ascending.
+// Only first event can be a snapshot.
+func (s *EventStore) Save(ctx context.Context, events ...eventstore.Event) (status eventstore.SaveStatus, err error) {
 	s.LogDebugfFunc("mongodb.Evenstore.Save start")
 	t := time.Now()
 	defer func() {
 		s.LogDebugfFunc("mongodb.Evenstore.Save takes %v", time.Since(t))
 	}()
 
-	if len(events) == 0 {
-		return false, errors.New("cannot save empty events")
-	}
-	if aggregateID == "" {
-		return false, errors.New("cannot save events without AggregateId")
+	if len(events) == 0 || events[0] == nil {
+		return eventstore.FAIL, fmt.Errorf("invalid events('%v')", events)
 	}
-
-	col := s.client.Database(s.DBName()).Collection(getEventCollectionName(collectionID))
-	if events[0].Version() == 0 {
-		err = s.ensureIndex(ctx, col, eventsQueryIndex)
-		if err != nil {
-			return false, fmt.Errorf("cannot save events: %w", err)
+	aggregateID := events[0].AggregateID()
+	groupID := events[0].GroupID()
+	version := events[0].Version()
+	for idx, event := range events {
+		if event == nil {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v]('%v')", idx, event)
+		}
+		if event.AggregateID() == "" {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].AggregateID('%v')", idx, event.AggregateID())
+		}
+		if event.GroupID() == "" {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].GroupID('%v')", idx, event.GroupID())
+		}
+		if event.EventType() == "" {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].EventType('%v')", idx, event.EventType())
+		}
+		if event.Version() != version+uint64(idx) {
+			return eventstore.FAIL, fmt.Errorf("invalid continues ascending events[%v].Version(%v))", idx, event.Version())
 		}
+		if idx > 0 && event.IsSnapshot() {
+			return eventstore.FAIL, fmt.Errorf("inner snapshot for events[%v] is not supported", idx)
+		}
+		if event.AggregateID() != aggregateID {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].AggregateID('%v') != events[0].AggregateID('%v')", idx, event.AggregateID(), aggregateID)
+		}
+		if event.GroupID() != groupID {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].GroupID('%v') != events[0].GroupID('%v')", idx, event.GroupID(), groupID)
+		}
+
 	}
 
+	col := s.client.Database(s.DBName()).Collection(getEventCollectionName())
 	if events[0].Version() == 0 {
-		concurrencyException, err = s.SaveSnapshotQuery(ctx, collectionID, aggregateID, 0)
+		doc, err := makeDBDoc(events, s.dataMarshaler)
 		if err != nil {
-			return false, fmt.Errorf("cannot save events without snapshot query for version 0: %w", err)
+			return eventstore.FAIL, fmt.Errorf("cannot insert first events('%v'): %w", events, err)
 		}
-		if concurrencyException {
-			return concurrencyException, nil
+		_, err = col.InsertOne(ctx, doc)
+		if err != nil {
+			if IsDup(err) {
```

Comment:
Is this still required even with a new mongo?

Reply:
Ofcourse it is needed.